pkg/e2e/spinup_test.go

```diff
@@ -0,0 +1,279 @@
+	"github.com/improbable-eng/kedge/protogen/e2e"
+	"github.com/oklog/run"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	httpWinchPort = "18070"
+	grpcWinchPort = "18071"
+
+	httpKedgePort    = "18170"
+	httpTLSKedgePort = "18171"
+	grpcTLSKedgePort = "18172"
+
+	httpTestEndpointPort = "18270"
+	grpcTestEndpointPort = "18271"
+
+	miscDir = "../../misc/"
+
+	endpointDNS       = "test_endpoint.localhost.internal.example.com"
+	noAuthEndpointDNS = "no_auth.test_endpoint.localhost.internal.example.com"
+)
+
+type config struct {
+	winch        bool
+	kedge        bool
+	testEndpoint bool
+}
+
+// NOTE: It is important to `make build` before using this function to compile latest changes.
+// spinup runs winch, kedge and test server, connected to each other.
+func spinup(t testing.TB, ctx context.Context, cfg config) (chan error, error) {
+	var commands []*exec.Cmd
+
+	if cfg.winch {
+		commands = append(commands, exec.Command("winch",
+			"--server_http_port", httpWinchPort,
+			"--server_grpc_port", grpcWinchPort,
+			"--server_mapper_config_path", miscDir+"winch_mapper.json",
+			"--server_auth_config_path", miscDir+"winch_auth.json",
+			"--pac_redirect_sh_expressions", "*.*.internal.example.com",
+			"--client_tls_cert_file", miscDir+"client.crt",
+			"--client_tls_key_file", miscDir+"client.key",
+			"--client_tls_root_ca_files", miscDir+"ca.crt",
+			"--debug_mode", "true",
+		))
+	}
+
+	if cfg.kedge {
+		commands = append(commands, exec.Command("kedge",
+			"--server_http_port", httpKedgePort,
+			"--server_http_tls_port", httpTLSKedgePort,
+			"--server_grpc_tls_port", grpcTLSKedgePort,
+			"--server_tls_cert_file", miscDir+"localhost.crt",
+			"--server_tls_key_file", miscDir+"localhost.key",
+			"--server_tls_client_ca_files", miscDir+"ca.crt",
+			"--server_tls_client_cert_required", "true",
+			"--kedge_config_director_config_path", miscDir+"director.json",
+			"--kedge_config_backendpool_config_path", miscDir+"backendpool.json",
+			"--log_level", "debug",
+		))
+	}
+
+	var (
+		g run.Group
+	)
+
+	if cfg.testEndpoint {
+		// Tokens are specified in misc/winch_auth.json
+		err := startTestEndpoints(&g, "secret-token2", "secret-token")
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Interrupt go routine.
+	{
+		ctx, cancel := context.WithCancel(ctx)
+		g.Add(func() error {
+			<-ctx.Done()
+
+			// This go routine will return only when:
+			// 1) Any other process from group exited unexpectedly
+			// 2) Global context will be cancelled.
+			return nil
+		}, func(error) {
+			cancel()
+		})
+	}
+
+	// Run go routine for each command.
+	for _, c := range commands {
+		var stderr, stdout bytes.Buffer
+		c.Stderr = &stderr
+		c.Stdout = &stdout
+
+		err := c.Start()
+		if err != nil {
+			// Let already started commands finish.
+			go g.Run()
+			return nil, errors.Wrap(err, "failed to start")
+		}
+
+		cmd := c
```

Comment:
This the classic Golang closure + loop problems. When you have closure and you passes some variable there.. it takes just "reference" of that var. Since the var changes in every loop (`c` here) you can end up with weird state where every `g.Add` is filled with last `c` item, not each one seperated (:

Reply:
ahhhh yes I have read about this before 👍  thanks for clarification.